### PR TITLE
Fix #5373

### LINKED ISSUE #5373: Enum completion prevention of "Color Color" ambiguity
**User Scenario**

In a WPF control with a nested control named `subControl`, type `subControl.Visibility = Visibility.Hidden` in the .xaml.cs file. Because `Visibility` is the name of an Enum and also a property of the outer control, completion after `Visibility.` contains enum members _and_ static/instance members of the Visibility property. 

Typing the same thing in Visual Studio 2013 prevents this problem by fully qualifying `Visibility` as `System.Windows.Visibility` when completed, removing the subsequent ambiguity.

**Standalone (non-WPF) Repro**

Copy/paste the following C# code:

``` C#
namespace S
{
    class C
    {
        Fail Fail;
        Pass Passes;

        void M()
        {
            C c = new C();
            c.Passes = Pass.A; // Completion offers "Pass" and then "A", as expected.
            c.Fail = Fail.A; // Completion offers "Fail" and then { "A", "CompareTo", "Equals", etc. }
                            // In VS2013, "Fail" on the RHS would complete as "S.FAIL" to prevent this.
        }
    }

    enum Fail { A };
    enum Pass { A };
}
```

When typing the `A` part of `c.Fail = Fail.A`, the completion list appears to contain both members of the Fail enum AND static/instance members of the Fail field.

Visual Studio 2013 attempts to detect and prevent this scenario in advance by completing `Fail` as `S.Fail`, which then causes only the enum members to be listed (as desired). If completion is invoked on an already-typed `Fail.`, then the behavior matches that of Visual Studio 2015. 

Applying the same fix in Visual Studio 2015 would be awkward because the `S.` qualification would immediately be marked as unnecessary. Perhaps there's some other way we can address this scenario.

### PR DESCRIPTION
**Pull request by rchande**
When showing completion after ., detect a "Color color" scenario between
an enum and a field with the same name. In that case, show only enum
members.

Tagging @dotnet/roslyn-ide for review.

### CODE REVIEW

**Comment by davkean** [comment]
(Closing and reopening to sync with jenkins)

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs [hunk 1]
```diff
@@ -2454,7 +2454,7 @@ void M()
     {
         $$";
 
-            await VerifyItemExistsAsync(markup, "M", expectedDescriptionOrNull: $"void C.M(int i) (+ 1 {FeaturesResources.Overload})");
+            await VerifyItemExistsAsync(markup, "M", expectedDescriptionOrNull: $"void C.M(int i) (+ 1 {FeaturesResources.Overload})");
```
Comment (rchande) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs]: Not sure why there are showing up as diffs.

**Comment by rchande** [comment]
Ping @dotnet/roslyn-ide? It's a pretty simple change.

File: src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 2]
```diff
@@ -2389,21 +2389,26 @@ public static bool IsEnumTypeMemberAccessContext(this SyntaxTree syntaxTree, Sem
...
-                return false;
+                return true;
             }
 
-            switch (symbol.Kind)
+            // Finally, check if it's an alias to an enum
+            var leftHandBinding = semanticModel.GetSymbolInfo(memberAccess.Expression, cancellationToken);
+            var symbol = leftHandBinding.GetBestOrAllSymbols().FirstOrDefault();
+
+            if (symbol != null && symbol.Kind == SymbolKind.Alias)
```
Comment (DustinCampbell) [inline src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs]:
```C#
if (symbol?.Kind == SymbolKind.Alias)
```

**Comment by DustinCampbell** [comment]
:+1:

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs (hunk 1 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs]: Did we mangle some nonbreaking space somehow, or maybe this is a GitHub bug? What does command line git say?

Comment (DustinCampbell) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs]: Great point -- this is probably a place where we had a non-breaking space.

**Comment by CyrusNajmabadi** [comment]
Wait
Why only show enum members?  If you have a field/type with the same name, i'd expect you to show the merged list...

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs (hunk 1 above)
Comment (rchande) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs]: If I take the raw file contents of this file from Master and paste them over the copy in my enlistment, Git shows no diff between what I committed and what I pasted, which makes me think this is a GitHub bug. Unless there's a clipboard issue?

Comment (rchande) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs]: Oh, but the tests with these diffs failed which suggests that something did actually get mangled.

File: src/EditorFeatures/VisualBasicTest/Completion/CompletionProviders/SymbolCompletionProviderTests.vb [hunk 3]
```diff
@@ -2509,7 +2509,7 @@ End Class
             Await VerifyItemExistsAsync(markup, "a")
             Await VerifyItemExistsAsync(markup, "b")
             Await VerifyItemExistsAsync(markup, "c")
-            Await VerifyItemIsAbsentAsync(markup, "Equals")
+            Await VerifyItemExistsAsync(markup, "Equals")
```
Comment (Pilchie) [inline src/EditorFeatures/VisualBasicTest/Completion/CompletionProviders/SymbolCompletionProviderTests.vb]: This part isn't right - We shouldn't show static members from `System.Enum` on instances of actual Enum types.

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationService.cs [hunk 4]
```diff
@@ -88,7 +88,20 @@ internal bool ShouldIncludeSymbol(ISymbol symbol)
 
                 if (_context.IsEnumTypeMemberAccessContext)
                 {
-                    return symbol.Kind == SymbolKind.Field;
+                    // If you have a field and an enum with the same name,
+                    // we want to show instance members (through the field)
+                    // so you can call CompareTo, etc, in addition to the 
+                    // enum fields. However, we don't want to show statics
+                    // from the Enum base type.
+                    if (symbol.Kind == SymbolKind.Field)
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationService.cs]: This seems weird.  Presumably properties are also affected.  (And locals/parameters if they happened to be the same casing as the enum?)

File: src/EditorFeatures/VisualBasicTest/Completion/CompletionProviders/SymbolCompletionProviderTests.vb (hunk 3 above)
Comment (rchande) [inline src/EditorFeatures/VisualBasicTest/Completion/CompletionProviders/SymbolCompletionProviderTests.vb]: This is an override of the instance method `Object.Equals`. Should we be explicitly filtering it out?

**Comment by DustinCampbell** [comment]
@rchande: What's the status of this PR?

**Pull request closed without merging**